text/0019-stricter-validation.md

```diff
@@ -0,0 +1,143 @@
+We will infer the type `False` for these expression if the slot is an entity type that cannot be `==` or `in` the principal or resource entity type, short-circuiting the rest of typechecking for that type environment.
+Typechecking could be optimized by recognizing and ignoring type environments that will be `False` based only on the policy scope.
+
+For example, using the schema from the motivating example, the following template will typecheck.
+
+```
+permit(principal == ?principal, action == Action::"read", resource);
+```
+
+Using the usual request type environment construction, there is one environment `(User, Action::"read", Object)`.
+Because this is a template containing `?principal`, we extend the environment with each possible type for the slot: `User`, `Org`, `Admin` and `Object`.
+When typechecking, the validator sees `principal == ?principal && ...` first in every environment which is always an equality between two entity types.
+The `principal` is always a `User` while `?principal` is one of the possible slot types depending on the type environment.
+In every environment where `?principal` is not `User`, the equality has type `False`, causing the `&&` to short-circuit to type `False`,
+The rest of the policy is typechecked as usual when `?principal` is `User`.
+
+This change to template typechecking does not introduce any new type errors and in fact will enable more precise typechecking if we expand where in a policy template slots may be used.
+
+### Unspecified principal/resource entity types
+
+The validator has a concept of an unspecified principal or resource entity type in an action `appliesTo` specification.
+When `principalTypes` or `resourceTypes` is omitted, we interpret the schema as declaring that the action does not apply to any particular principal/resource entity type.
+It instead applies to the "unspecified" principal or resource, which allows a users to make authorization requests using that action without providing the principal or resource,
+or when providing an entity with any of the defined entity types as the principal or resource.
+This is implemented by assigning the `AnyEntity` type to a variable when it is unspecified,
+but this depends on width subtyping between entity types which we want to eliminate from strict validation.
+
+It is not possible to extend the set of type environments to handle this case in a way that does not use `AnyEntity`.
+Constructing one type environment for each declared entity type does not soundly handle the case where an entity is _not_ provided in the query.
+If some attribute exists for every declared entity type, then an access to that attribute would be permitted, but that attribute cannot be safely accessed if an entity is not used in the query.
+We would need to further extend set of type environments to include an entity representing this case,
+but, other than `AnyEntity`, we do not have a type that can represent the type of an unspecified entity.
```

Comment:
What is special about `AnyEntity` type here?  Or, put differently, what is preventing us from constructing a synthetic entity type with no attributes, say `ArbitraryEntity`, and using that instead of `AnyEntity`?  Or, maybe this text is saying that such a solution is equivalent to using `AnyEntity` without subtyping, and we're simply reusing the name `AnyEntity` from permissive validation?

Reply:
`AnyEntity` isn't special. The point is that if strict mode doesn't allow for any subtyping with `AnyEntity`, then `AnyEntity` becomes this arbitrary entity type.

Reply:
Got it, so we're reusing the name `AnyEntity` but not any of the associated properties from permissive validation.  In that case, maybe we want to give it a different name in the context of strict validation to avoid confusion.  Maybe something like `ArbitraryEntity`.